src/librustc_trans/back/link.rs

```diff
@@ -508,23 +563,548 @@ fn link_binary_output(sess: &Session,
+    };
+    if force_non_stable_output || sess.opts.debuginfo != config::NoDebugInfo {
+        // emit bc for translation into nexe w/ debugging info.
+        let outpath = outputs.with_extension("debug.pexe");
+        let out = outpath.display().to_string();
+        check_writeable_output(sess, &outpath, "debug output");
+        let cout = format!("{}\0", out);
+        unsafe {
+            llvm::LLVMWriteBitcodeToFile(llmod, cout.as_ptr() as *const i8)
+        };
+        set_permissions(&outpath);
+    }
+
+    let emit_stable_pexe = true; // always emit stable bitcode.
+    if emit_stable_pexe {
+        unsafe {
+            // The PNaCl stable bitcode format doesn't accept metadata types.
+            llvm::LLVMRustStripDebugInfo(llmod);
+
+            let pm = llvm::LLVMCreatePassManager();
+
+            let ap = |s: &'static str| {
+                assert!(llvm::LLVMRustAddPass(pm, s.as_ptr() as *const i8),
+                        "failed to add pass `{}`", &s[0 .. s.len() - 1]);
+            };
+
+            // Strip unsupported metadata and things:
+            ap("strip-metadata\0");
+            ap("strip-module-flags\0");
+            ap("nacl-strip-attributes\0");
+
+            if !sess.no_verify() {
+                ap("verify-pnaclabi-module\0");
+                ap("verify-pnaclabi-functions\0");
+            }
+
+            llvm::LLVMRunPassManager(pm, llmod);
+
+            llvm::LLVMDisposePassManager(pm);
+
+            llvm::LLVMRustStripDebugInfo(llmod);
+        }
+    }
+
+    let out = match outputs.single_output_file {
+        Some(ref file) => file.clone(),
+        None => {
+            let out_filename = outputs.path(OutputTypeExe);
+            filename_for_input(sess, config::CrateTypeExecutable, crate_name, &out_filename)
+        }
+    };
+    let out_cstr = format!("{}\0", out.display().to_string());
+
+    check_writeable_output(sess, &out, "final output");
+
+    if emit_stable_pexe {
+        if !unsafe { llvm::LLVMRustWritePNaClBitcode(llmod,
+                                                     out_cstr.as_ptr() as *const i8,
+                                                     false) } {
+            llvm_err(&sess.diagnostic().handler, "failed to write output file".to_string());
+        }
+    } else {
+        // regular bitcode output:
+        unsafe { llvm::LLVMWriteBitcodeToFile(llmod, out_cstr.as_ptr() as *const i8) };
+    }
+
+
+    #[cfg(unix)]
+    fn set_permissions(out: &PathBuf) {
+        use std::os::unix::fs::PermissionsExt;
+        use std::fs::{metadata, set_permissions};
+
+        let mut perms = metadata(out)
+            .unwrap()
+            .permissions();
+        perms.set_mode(0o0755);
+        set_permissions(out, perms)
+            .unwrap();
+    }
+    #[cfg(not(unix))]
+    fn set_permissions(_: &PathBuf) { }
+
+    set_permissions(&out);
+
+    unsafe {
+        llvm::LLVMContextDispose(llcx);
+    }
```

Comment:
This is another massive amount of plattform-specific code that other readers are going to be clueless about and not want to see. There must be abstractions we can rationalize here to hide the pnacle stuff behind.